sparkly/functions.py

```diff
@@ -153,3 +153,45 @@ def _execute_case(accumulator, case):
     result = reduce(_execute_case, cases, F).otherwise(default)
 
     return result
+
+
+def argmax(field, by, condition=None, nullable=False):
+    """Select a value from the row that maximizes other column(s)
+
+    Arguments:
+        field (string, pyspark.sql.Column): the field to return (and the alias to use). If type
+            is a string, the resulting column would have the same name; otherwise, it would be
+            named '__tmp_argmax__' and needs to be renamed as needed
+        nullable (boolean): whether the field can be None
+        by (*string, *pyspark.sql.Column): field or list of fields to maximize. In reality, this
+            will usually be only one field. But you may use multiple for tiebreakers
+        condition (optional): Only consider the entities that pass this condition
+
+    Example:
+        df = (
+            df
+            .groupBy('id')
+            .agg(argmax('field1', 'by_field', nullable=True))
+        )
+
+        argmax('field1', ['by_field1', 'by_field2'], condition=F.col('col') == 1)
+        argmax(F.col('field1'), [F.col('by_field1'), F.col('by_field2')], condition=F.lit(True))
+    """
+    if not isinstance(by, list):
+        by = [by]
+
+    col = F.col(field) if isinstance(field, str) else field
+
+    by.append(col.alias('__tmp_argmax__'))
```

Comment:
I think we can do this without fake-name hacks, (see implementation of similar functionality in different places). If you really think it's a more generic solution than those let's add justification here or in code comments

Reply:
Here is the reasoning I have for this:
1. We need to know the name since the getField function call later on. There is another way of using df.col._jc.toString().encode('utf8') to get the column name get passed in. Do you prefer to do it that way?
2. It will have a fixed name for the resulting column, if the input 'by' variable is a complex col expression, the resulting column name would be whatever that expression is.